internal/handler/msg_usersinfo.go

```diff
@@ -29,6 +35,248 @@ func (h *Handler) MsgUsersInfo(ctx context.Context, msg *wire.OpMsg) (*wire.OpMs
+	case *types.Array:
+		for i := 0; i < user.Len(); i++ {
+			var ui any
+			ui, err = user.Get(i)
+
+			if err != nil {
+				return nil, lazyerrors.Error(err)
+			}
+
+			if ui != nil {
+				var u usersInfoPair
+
+				if err = u.extract(ui, dbName); err != nil {
+					return nil, lazyerrors.Error(err)
+				}
+
+				users = append(users, u)
+			}
+		}
+	case string:
+		var u usersInfoPair
+		if err = u.extract(user, dbName); err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		users = append(users, u)
+	case int32: // {usersInfo: 1 }
+		singleDB = true
+	default:
+		return nil, handlererrors.NewCommandErrorMsgWithArgument(
+			handlererrors.ErrBadValue,
+			"UserName must be either a string or an object",
+			document.Command(),
+		)
+	}
+
+	// Users are saved in the "admin" database.
+	adminDB, err := h.b.Database("admin")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	usersCol, err := adminDB.Collection("system.users")
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	var filter *types.Document
+	filter, err = usersInfoFilter(allDBs, singleDB, dbName, users)
+
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	// Filter isn't being passed to the query as we are filtering after retrieving all data
+	// from the database due to limitations of the internal/backends filters.
+	qr, err := usersCol.Query(ctx, nil)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	defer qr.Iter.Close()
+
+	var res *types.Array
+	res, err = types.NewArray()
+
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	for {
+		_, v, err := qr.Iter.Next()
+
+		if errors.Is(err, iterator.ErrIteratorDone) {
+			break
+		}
+
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		matches, err := common.FilterDocument(v, filter)
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+
+		if matches {
+			res.Append(v)
+		}
+	}
+
+	var reply wire.OpMsg
+	must.NoError(reply.SetSections(wire.OpMsgSection{
+		Documents: []*types.Document{must.NotFail(types.NewDocument(
+			"users", res,
+			"ok", float64(1),
+		))},
+	}))
+
+	return &reply, nil
+}
+
+// usersInfoPair is a pair of username and database name.
+type usersInfoPair struct {
+	username string
+	db       string
+}
+
+// extract username and db from v.
+func (p *usersInfoPair) extract(v any, dbName string) error {
+	switch vt := v.(type) {
+	case *types.Document:
+		ui, err := vt.Get("user")
+		if err != nil {
+			return lazyerrors.Error(err)
+		}
+
+		var ok bool
+		p.username, ok = ui.(string)
+
+		if !ok {
+			return handlererrors.NewCommandErrorMsg(
+				handlererrors.ErrBadValue,
+				fmt.Sprintf("UserName must contain a string field named: user. But, has type %T", ui),
+			)
+		}
+
+		if !vt.Has("db") {
+			return handlererrors.NewCommandErrorMsg(
+				handlererrors.ErrBadValue,
+				"UserName must contain a field named: db",
+			)
+		}
+
+		db, err := vt.Get("db")
+		if err != nil {
+			return lazyerrors.Error(err)
+		}
```

Comment:
I think this can be simplified a bit. `vt.Get` will return error in a single case - if `db` field doesn't exist. So, no need to have an extra check.

```suggestion
		db, err := vt.Get("db")
		if err != nil {
		     return handlererrors.NewCommandErrorMsg(
				handlererrors.ErrBadValue,
				"UserName must contain a field named: db",
			)
		}
```